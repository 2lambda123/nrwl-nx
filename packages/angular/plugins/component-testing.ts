import {
  ProjectGraph,
  createProjectGraphAsync,
  ProjectGraphDependency,
} from '@nrwl/devkit';
import { nxBaseCypressPreset } from '@nrwl/cypress/plugins/cypress-preset';

export function nxComponentTestingPreset(pathToConfig: string) {
  return {
    ...nxBaseCypressPreset(pathToConfig),
    devServer: {
      // cypress uses string union type,
      // need to use const to prevent typing to string
      framework: 'angular',
      bundler: 'webpack',
      // TODO(caleb): use project graph to get the correct ng config build path.
    } as const,
  };
}

async function getBuildConfig(projectPath: string) {
  console.log('creating graph');
  const pg = await createProjectGraphAsync();
  let projectName: string;

  for (const n in pg.nodes) {
    console.log(pg.nodes[n].data.root);
    if (pg.nodes[n].data.root === projectPath) {
      console.log('found name', pg.nodes[n].name);
      projectName = pg.nodes[n].name;
    }
  }

  if (!projectName) {
    console.log('oops didnt find project name in path', projectPath);
    return;
  }

  const project = pg.nodes[projectName];
  console.log('looking for build target');

  if (project.data.targets['build']) {
    console.log('project has build target', project.data.targets['build']);
  } else {
    const parents = findParentsOfProject(pg, projectName);
    if (parents.length === 0) {
      console.log('no build targets found in parents of project', projectName);
    } else {
      console.log('found build targets in parents of project', projectName);

      for (const parent of parents) {
        // TODO(caleb): how do I tell which is the "correct" one to use? i.e. which is the "closest" to our projectName?
        console.log(parent.source, pg.nodes[parent.source].data.targets);
      }
    }
  }
}

function findParentsOfProject(
  pg: ProjectGraph,
  projectName: string
): ProjectGraphDependency[] {
  const parents = [];
  for (const dep in pg.dependencies) {
    const f = pg.dependencies[dep].filter((d) => d.target === projectName);
    parents.push(...f);
  }
  return parents;
}
