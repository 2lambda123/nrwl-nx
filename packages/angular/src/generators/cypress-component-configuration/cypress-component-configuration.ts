import { cypressComponentProject } from '@nrwl/cypress';
import {
  formatFiles,
  generateFiles,
  joinPathFragments,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
} from '@nrwl/devkit';
import { componentTestGenerator } from '../component-test/component-test';
import {
  getComponentsInfo,
  getStandaloneComponentsInfo,
} from '../utils/storybook-ast/component-info';
import { getProjectEntryPoints } from '../utils/storybook-ast/entry-point';
import { getModuleFilePaths } from '../utils/storybook-ast/module-info';
import { CypressComponentConfigSchema } from './schema';

/**
 * This is for cypress built in component testing, if you want to test with
 * storybook + cypress then use the componentCypressGenerator instead.
 */
export async function cypressComponentConfiguration(
  tree: Tree,
  options: CypressComponentConfigSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const installTask = await cypressComponentProject(tree, {
    project: options.project,
    skipFormat: true,
  });

  addFiles(tree, projectConfig, options);

  if (options.skipFormat) {
    await formatFiles(tree);
  }
  return () => {
    installTask();
  };
}
function addFiles(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  options: CypressComponentConfigSchema
) {
  const cypressConfigPath = joinPathFragments(
    joinPathFragments(projectConfig.root, 'cypress.config.ts')
  );

  if (tree.exists(cypressConfigPath)) {
    tree.delete(cypressConfigPath);
  }
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    projectConfig.root,
    {
      tpl: '',
    }
  );

  if (options.generateTests) {
    const entryPoints = getProjectEntryPoints(tree, options.project);

    const componentInfo = [];
    for (const entryPoint of entryPoints) {
      const moduleFilePaths = getModuleFilePaths(tree, entryPoint);
      componentInfo.push(
        ...getComponentsInfo(
          tree,
          entryPoint,
          moduleFilePaths,
          options.project
        ),
        ...getStandaloneComponentsInfo(tree, entryPoint)
      );
    }

    for (const info of componentInfo) {
      if (info === undefined) {
        continue;
      }
      componentTestGenerator(tree, {
        projectPath: info.moduleFolderPath,
        componentName: info.name,
        componentDir: info.path,
        componentFileName: info.componentFileName,
        skipFormat: true,
      });
    }
  }
}
