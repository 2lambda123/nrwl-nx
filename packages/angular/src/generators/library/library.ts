import {
  formatFiles,
  installPackagesTask,
  moveFilesToNewDirectory,
  Tree,
} from '@nrwl/devkit';
import { wrapAngularDevkitSchematic } from '@nrwl/devkit/ngcli-adapter';
import { jestProjectGenerator } from '@nrwl/jest';
import { Linter } from '@nrwl/workspace';

import init from '../../generators/init/init';
import addLintingGenerator from '../add-linting/add-linting';
import karmaProjectGenerator from '../karma-project/karma-project';

import { addModule } from './lib/add-module';
import { normalizeOptions } from './lib/normalize-options';
import { updateLibPackageNpmScope } from './lib/update-lib-package-npm-scope';
import { updateProject } from './lib/update-project';
import { updateTsConfig } from './lib/update-tsconfig';
import {
  enableStrictTypeChecking,
  setLibraryStrictDefault,
} from './lib/enable-strict-type-checking';
import { NormalizedSchema } from './lib/normalized-schema';
import { Schema } from './schema';

export default async function (host: Tree, schema: Schema) {
  // Do some validation checks
  const options = normalizeOptions(host, schema);
  if (!options.routing && options.lazy) {
    throw new Error(`routing must be set`);
  }

  if (options.enableIvy === true && !options.buildable) {
    throw new Error('enableIvy must only be used with buildable.');
  }

  if (options.publishable === true && !schema.importPath) {
    throw new Error(
      `For publishable libs you have to provide a proper "--importPath" which needs to be a valid npm package name (e.g. my-awesome-lib or @myorg/my-lib)`
    );
  }

  console.log('Initializing library generator');
  init(host, {
    ...options,
    skipFormat: true,
  });

  console.log('Scaffolding initial files');
  const runAngularLibrarySchematic = wrapAngularDevkitSchematic(
    '@schematics/angular',
    'library'
  );
  await runAngularLibrarySchematic(host, {
    name: options.name,
    prefix: options.prefix,
    entryFile: 'index',
    skipPackageJson: !(options.publishable || options.buildable),
    skipTsConfig: false,
  });

  console.log('Moving to correct directory');
  moveFilesToNewDirectory(host, options.name, options.projectRoot);

  console.log('Updating project configuration');
  await updateProject(host, options);

  console.log("Updating project's TS Config");
  updateTsConfig(host, options);

  console.log('Adding unit test runner');
  await addUnitTestRunner(host, options);

  console.log('If Buildable/Publishable, fix up npm scope');
  updateNpmScopeIfBuildableOrPublishable(host, options);

  console.log('Add Angular Module');
  addModule(host, options);

  console.log('Determine if strict mode should be set');
  setStrictMode(host, options);

  console.log('Add Linter');
  await addLinting(host, options);

  console.log('Formatting files');
  await formatFiles(host);

  return () => {
    installPackagesTask(host);
  };
}

async function addUnitTestRunner(host: Tree, options: NormalizedSchema) {
  if (options.unitTestRunner === 'jest') {
    await jestProjectGenerator(host, {
      project: options.name,
      setupFile: 'angular',
      supportTsx: false,
      skipSerializers: false,
    });
  } else if (options.unitTestRunner === 'karma') {
    await karmaProjectGenerator(host, {
      project: options.name,
    });
  }
}

function updateNpmScopeIfBuildableOrPublishable(
  host: Tree,
  options: NormalizedSchema
) {
  if (options.buildable || options.publishable) {
    updateLibPackageNpmScope(host, options);
  }
}

function setStrictMode(host: Tree, options: NormalizedSchema) {
  if (options.strict) {
    enableStrictTypeChecking(host, options);
  } else {
    setLibraryStrictDefault(host, options.strict);
  }
}

async function addLinting(host: Tree, options: NormalizedSchema) {
  if (options.linter === Linter.None) {
    return;
  }
  await addLintingGenerator(host, {
    projectName: options.name,
    projectRoot: options.projectRoot,
    prefix: options.prefix,
  });
}
