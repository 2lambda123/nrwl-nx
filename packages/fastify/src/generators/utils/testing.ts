import { Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';

export function createTreeWithFastifyApplication(appName: string): Tree {
  const tree = createTreeWithEmptyWorkspace();
  tree.write(
    'workspace.json',
    String.raw`
      {
        "projects": {
          "${appName}": {
            "root": "apps/api",
            "sourceRoot": "apps/api/src",
            "projectType": "application",
            "prefix": "api",
            "targets":{}
          }
        }
      }
    `
  );

  return tree;
}
