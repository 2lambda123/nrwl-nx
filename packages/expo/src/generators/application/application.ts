import {
  formatFiles,
  GeneratorCallback,
  joinPathFragments,
  runTasksInSerial,
  Tree,
} from '@nx/devkit';
import { initGenerator as jsInitGenerator } from '@nx/js';

import { runSymlink } from '../../utils/symlink-task';
import { addLinting } from '../../utils/add-linting';
import { addJest } from '../../utils/add-jest';

import { normalizeOptions } from './lib/normalize-options';
import initGenerator from '../init/init';
import { addProject } from './lib/add-project';
import { createApplicationFiles } from './lib/create-application-files';
import { addEasScripts } from './lib/add-eas-scripts';
import { addDetox } from './lib/add-detox';
import { Schema } from './schema';
import { ensureDependencies } from '../../utils/ensure-dependencies';
import { initRootBabelConfig } from '../../utils/init-root-babel-config';
import { logShowProjectCommand } from '@nx/devkit/src/utils/log-show-project-command';

export async function expoApplicationGenerator(
  host: Tree,
  schema: Schema
): Promise<GeneratorCallback> {
  return await expoApplicationGeneratorInternal(host, {
    addPlugin: false,
    projectNameAndRootFormat: 'derived',
    ...schema,
  });
}

export async function expoApplicationGeneratorInternal(
  host: Tree,
  schema: Schema
): Promise<GeneratorCallback> {
  const options = await normalizeOptions(host, schema);

  const tasks: GeneratorCallback[] = [];
  const jsInitTask = await jsInitGenerator(host, {
    ...schema,
    skipFormat: true,
  });
  tasks.push(jsInitTask);
  const initTask = await initGenerator(host, { ...options, skipFormat: true });
  tasks.push(initTask);
  if (!options.skipPackageJson) {
    tasks.push(ensureDependencies(host));
  }
  initRootBabelConfig(host);

  createApplicationFiles(host, options);
  addProject(host, options);

  const lintTask = await addLinting(host, {
    ...options,
    projectRoot: options.appProjectRoot,
    tsConfigPaths: [
      joinPathFragments(options.appProjectRoot, 'tsconfig.app.json'),
    ],
  });
  tasks.push(lintTask);

  const jestTask = await addJest(
    host,
    options.unitTestRunner,
    options.projectName,
    options.appProjectRoot,
    options.js,
    options.skipPackageJson,
    options.addPlugin
  );
  tasks.push(jestTask);
  const detoxTask = await addDetox(host, options);
  tasks.push(detoxTask);
  const symlinkTask = runSymlink(host.root, options.appProjectRoot);
  tasks.push(symlinkTask);
  addEasScripts(host);

  if (!options.skipFormat) {
    await formatFiles(host);
  }

  tasks.push(() => {
    logShowProjectCommand(options.projectName);
  });

  return runTasksInSerial(...tasks);
}

export default expoApplicationGenerator;
