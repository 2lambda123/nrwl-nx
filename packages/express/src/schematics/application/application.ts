import {
  chain,
  externalSchematic,
  Rule,
  SchematicContext,
  Tree,
  noop,
} from '@angular-devkit/schematics';
import { join, normalize, Path } from '@angular-devkit/core';
import { updateJsonInTree, toFileName, formatFiles } from '@nrwl/workspace';
import { toJS } from '@nrwl/workspace/src/utils/rules/to-js';
import { appsDir } from '@nrwl/workspace/src/utils/ast-utils';
import init from '../init/init';
import { Schema } from './schema';
import { mainModule } from 'process';

interface NormalizedSchema extends Schema {
  appProjectRoot: Path;
}

function addTypes(options: NormalizedSchema): Rule {
  const tsConfigPath = join(options.appProjectRoot, 'tsconfig.app.json');
  return updateJsonInTree(tsConfigPath, (json) => {
    json.compilerOptions.types = [...json.compilerOptions.types, 'express'];
    return json;
  });
}

function addMainFile(options: NormalizedSchema): Rule {
  return (host: Tree) => {
    host.overwrite(
      join(options.appProjectRoot, options.js ? 'src/main.js' : 'src/main.ts'),
      `/**
 * This is not a production server yet!
 * This is only a minimal backend to get started.
 */

import * as express from 'express';

const app = express();

app.get('/api', (req, res) => {
  res.send({ message: 'Welcome to ${options.name}!' });
});

const port = process.env.port || 3333;
const server = app.listen(port, () => {
  console.log(\`Listening at http://localhost:\${port}/api\`);
});
server.on('error', console.error);
`
    );
  };
}

export default function (schema: Schema): Rule {
  return (host: Tree, context: SchematicContext) => {
    const options = normalizeOptions(host, schema);
    return chain([
      init({ ...options, skipFormat: true }),
      externalSchematic('@nrwl/node', 'application', schema),
      addMainFile(options),
      addTypes(options),
      options.js ? toJS() : noop(),
      formatFiles(options),
    ])(host, context);
  };
}

function normalizeOptions(host: Tree, options: Schema): NormalizedSchema {
  const appDirectory = options.directory
    ? `${toFileName(options.directory)}/${toFileName(options.name)}`
    : toFileName(options.name);
  const appProjectRoot = join(normalize(appsDir(host)), appDirectory);

  return {
    ...options,
    appProjectRoot,
  };
}
