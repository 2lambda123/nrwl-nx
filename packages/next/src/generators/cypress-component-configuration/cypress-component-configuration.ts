import {
  ensurePackage,
  formatFiles,
  generateFiles,
  GeneratorCallback,
  joinPathFragments,
  ProjectConfiguration,
  readProjectConfiguration,
  runTasksInSerial,
  Tree,
  updateProjectConfiguration,
  visitNotIgnoredFiles,
} from '@nx/devkit';
import { isComponent } from '@nx/react/src/utils/ct-utils';
import { CypressComponentConfigurationGeneratorSchema } from './schema';
import { nxVersion } from '../../utils/versions';
import { componentTestGenerator } from '@nx/react';
import { normalize, relative } from 'path';

export async function cypressComponentConfiguration(
  tree: Tree,
  options: CypressComponentConfigurationGeneratorSchema
) {
  const tasks: GeneratorCallback[] = [];

  const { cypressComponentProject } = ensurePackage<
    typeof import('@nx/cypress')
  >('@nx/cypress', nxVersion);
  tasks.push(
    await cypressComponentProject(tree, {
      project: options.project,
      skipFormat: true,
    })
  );
  const projectConfig = readProjectConfiguration(tree, options.project);

  projectConfig.targets['component-test'].options = {
    ...projectConfig.targets['component-test'].options,
    skipServe: true,
  };
  updateProjectConfiguration(tree, options.project, projectConfig);

  await addFiles(tree, projectConfig, options);

  const { webpackInitGenerator } = ensurePackage<typeof import('@nx/webpack')>(
    '@nx/webpack',
    nxVersion
  );
  tasks.push(
    await webpackInitGenerator(tree, {
      compiler: 'swc',
      uiFramework: 'react',
      skipFormat: true,
    })
  );

  if (!options.skipFormat) {
    await formatFiles(tree);
  }

  return runTasksInSerial(...tasks);
}

async function addFiles(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  opts: CypressComponentConfigurationGeneratorSchema
) {
  const cypressConfigPath = joinPathFragments(
    projectConfig.root,
    'cypress.config.ts'
  );
  if (tree.exists(cypressConfigPath)) {
    tree.delete(cypressConfigPath);
  }

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    projectConfig.root,
    {
      tailwind: ['js', 'cjs'].some((ext) =>
        tree.exists(
          joinPathFragments(projectConfig.root, `tailwind.config.${ext}`)
        )
      ),
      tpl: '',
    }
  );

  const ctFile = joinPathFragments(
    projectConfig.root,
    'cypress',
    'support',
    'component.ts'
  );

  tree.write(ctFile, `${tree.read(ctFile, 'utf-8')}import './styles.ct.css';`);

  if (opts.generateTests) {
    const filePaths = [];
    visitNotIgnoredFiles(tree, projectConfig.sourceRoot, (filePath) => {
      const fromProjectRootPath = relative(projectConfig.root, filePath);
      console.log({ fromProjectRootPath, filePath });
      // we don't generate tests for pages/server-side/appDir components
      if (
        fromProjectRootPath.includes('pages') ||
        fromProjectRootPath.includes('server') ||
        fromProjectRootPath.includes('app')
      ) {
        return;
      }

      if (isComponent(tree, filePath)) {
        filePaths.push(filePath);
      }
    });

    for (const filePath of filePaths) {
      await componentTestGenerator(tree, {
        project: opts.project,
        componentPath: filePath,
      });
    }
  }
}

export default cypressComponentConfiguration;
