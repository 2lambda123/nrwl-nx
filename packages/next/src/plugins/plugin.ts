import {
  CreateDependencies,
  CreateNodes,
  CreateNodesContext,
  NxJsonConfiguration,
  TargetConfiguration,
  detectPackageManager,
  readJsonFile,
  writeJsonFile,
} from '@nx/devkit';
import { dirname, join } from 'path';

import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { existsSync, readdirSync } from 'fs';

import { projectGraphCacheDirectory } from 'nx/src/utils/cache-directory';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { type NextConfig } from 'next';
import { PHASE_PRODUCTION_BUILD } from 'next/constants';
import { getLockFileName } from '@nx/js';

export interface NextPluginOptions {
  buildTargetName?: string;
  devTargetName?: string;
  startTargetName?: string;
}

const cachePath = join(projectGraphCacheDirectory, 'next.hash');
const targetsCache = existsSync(cachePath) ? readTargetsCache() : {};

const calculatedTargets: Record<
  string,
  Record<string, TargetConfiguration>
> = {};

function readTargetsCache(): Record<
  string,
  Record<string, TargetConfiguration>
> {
  return readJsonFile(cachePath);
}

function writeTargetsToCache(
  targets: Record<string, Record<string, TargetConfiguration>>
) {
  writeJsonFile(cachePath, targets);
}

export const createDependencies: CreateDependencies = () => {
  writeTargetsToCache(calculatedTargets);
  return [];
};

// TODO(nicholas): Add support for .mjs files
export const createNodes: CreateNodes<NextPluginOptions> = [
  '**/next.config.{js, cjs}',
  async (configFilePath, options, context) => {
    const projectRoot = dirname(configFilePath);

    // Do not create a project if package.json and project.json isn't there.
    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
    if (
      !siblingFiles.includes('package.json') &&
      !siblingFiles.includes('project.json')
    ) {
      return {};
    }

    options = normalizeOptions(options);

    const hash = calculateHashForCreateNodes(projectRoot, options, context, [
      getLockFileName(detectPackageManager(context.workspaceRoot)),
    ]);

    const targets =
      targetsCache[hash] ??
      (await buildNextTargets(configFilePath, projectRoot, options, context));

    calculatedTargets[hash] = targets;

    return {
      projects: {
        [projectRoot]: {
          root: projectRoot,
          targets,
        },
      },
    };
  },
];

async function buildNextTargets(
  nextConfigPath: string,
  projectRoot: string,
  options: NextPluginOptions,
  context: CreateNodesContext
) {
  const nextConfig = getNextConfig(nextConfigPath, context);
  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<string, TargetConfiguration> = {};

  targets[options.buildTargetName] = await getBuildTargetConfig(
    namedInputs,
    projectRoot,
    nextConfig
  );

  targets[options.devTargetName] = getDevTargetConfig(projectRoot);

  targets[options.startTargetName] = getStartTargetConfig(options, projectRoot);
  return targets;
}

async function getBuildTargetConfig(
  namedInputs: { [inputName: string]: any[] },
  projectRoot: string,
  nextConfig: NextConfig
) {
  const nextOutputPath = await getOutputs(projectRoot, nextConfig);
  // Set output path here so that `withNx` can pick it up.
  const targetConfig: TargetConfiguration = {
    command: `next build`,
    options: {
      cwd: projectRoot,
    },
    dependsOn: ['^build'],
    cache: true,
    inputs: getInputs(namedInputs),
    outputs: [nextOutputPath, `${nextOutputPath}/!(cache)`],
  };
  return targetConfig;
}

function getDevTargetConfig(projectRoot: string) {
  const targetConfig: TargetConfiguration = {
    command: `next dev`,
    options: {
      cwd: projectRoot,
    },
  };

  return targetConfig;
}

function getStartTargetConfig(options: NextPluginOptions, projectRoot: string) {
  const targetConfig: TargetConfiguration = {
    command: `next start`,
    options: {
      cwd: projectRoot,
    },
    dependsOn: [options.buildTargetName],
  };

  return targetConfig;
}

async function getOutputs(projectRoot, nextConfig) {
  let dir = '.next';

  if (typeof nextConfig === 'function') {
    // Works for both async and sync functions.
    const configResult = await Promise.resolve(
      nextConfig(PHASE_PRODUCTION_BUILD, { defaultConfig: {} })
    );
    if (configResult?.distDir) {
      dir = configResult?.distDir;
    }
  } else if (typeof nextConfig === 'object' && nextConfig?.distDir) {
    // If nextConfig is an object, directly use its 'distDir' property.
    dir = nextConfig.distDir;
  }
  return `{workspaceRoot}/${projectRoot}/${dir}`;
}

function getNextConfig(
  configFilePath: string,
  context: CreateNodesContext
): Promise<any> {
  const resolvedPath = join(context.workspaceRoot, configFilePath);

  const module = load(resolvedPath);
  return module.default ?? module;
}

function normalizeOptions(options: NextPluginOptions): NextPluginOptions {
  options ??= {};
  options.buildTargetName ??= 'build';
  options.devTargetName ??= 'dev';
  options.startTargetName ??= 'start';
  return options;
}

function getInputs(
  namedInputs: NxJsonConfiguration['namedInputs']
): TargetConfiguration['inputs'] {
  return [
    ...('production' in namedInputs
      ? ['default', '^production']
      : ['default', '^default']),
    {
      externalDependencies: ['next'],
    },
  ];
}

/**
 * Load the module after ensuring that the require cache is cleared.
 */
function load(path: string): any {
  // Clear cache if the path is in the cache
  if (require.cache[path]) {
    for (const k of Object.keys(require.cache)) {
      delete require.cache[k];
    }
  }

  // Then require
  return require(path);
}
