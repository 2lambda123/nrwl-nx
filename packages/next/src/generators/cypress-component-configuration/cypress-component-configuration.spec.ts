import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { readJson, readProjectConfiguration, Tree } from '@nx/devkit';
import { cypressComponentConfiguration } from './cypress-component-configuration';
import { applicationGenerator } from '../application/application';
import { libraryGenerator } from '../library/library';
import { setupTailwindGenerator } from '@nx/react';
import { Linter } from '@nx/linter';

describe('cypress-component-configuration generator', () => {
  let tree: any;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  });

  it('should setup nextjs app', async () => {
    await applicationGenerator(tree, { name: 'demo', style: 'css' });
    await cypressComponentConfiguration(tree, {
      generateTests: true,
      project: 'demo',
    });
    expect(readJson(tree, 'apps/demo/tsconfig.json').exclude).toEqual(
      expect.arrayContaining([
        'cypress/**/*',
        'cypress.config.ts',
        '**/*.cy.ts',
        '**/*.cy.js',
        '**/*.cy.tsx',
        '**/*.cy.jsx',
      ])
    );
    expect(
      readJson(tree, 'apps/demo/cypress/tsconfig.cy.json')
    ).toMatchSnapshot();
    expect(tree.read('apps/demo/cypress.config.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { defineConfig } from 'cypress';
      import { nxComponentTestingPreset } from '@nx/next/plugins/component-testing';

      export default defineConfig({
        component: nxComponentTestingPreset(__filename),
      });
      "
    `);
    expect(tree.exists('apps/demo/pages/index.cy.ts')).toBeFalsy();
    expect(
      readProjectConfiguration(tree, 'demo').targets['component-test']
    ).toEqual({
      executor: '@nx/cypress:cypress',
      options: {
        cypressConfig: 'apps/demo/cypress.config.ts',
        skipServe: true,
        testingType: 'component',
      },
    });
  });

  it('should add styles setup in app', async () => {
    await applicationGenerator(tree, { name: 'demo', style: 'css' });
    await setupTailwindGenerator(tree, { project: 'demo' });
    await cypressComponentConfiguration(tree, {
      generateTests: false,
      project: 'demo',
    });

    expect(tree.read('apps/demo/cypress/support/styles.ct.css', 'utf-8'))
      .toMatchInlineSnapshot(`
      "/* This is where you can load global styles to apply to all components. */
      @tailwind base;
      @tailwind components;
      @tailwind utilities;
      "
    `);
    expect(tree.read('apps/demo/cypress/support/component.ts', 'utf-8'))
      .toContain(`import './commands';
import './styles.ct.css';`);
  });

  it('should setup nextjs lib', async () => {
    await libraryGenerator(tree, {
      name: 'demo',
      linter: Linter.EsLint,
      style: 'css',
      unitTestRunner: 'jest',
      component: true,
    });
    await cypressComponentConfiguration(tree, {
      generateTests: true,
      project: 'demo',
    });
    expect(readJson(tree, 'libs/demo/tsconfig.json').references).toEqual(
      expect.arrayContaining([
        {
          path: './cypress/tsconfig.cy.json',
        },
      ])
    );
    expect(
      readJson(tree, 'libs/demo/cypress/tsconfig.cy.json')
    ).toMatchSnapshot();
    expect(tree.read('libs/demo/cypress.config.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { defineConfig } from 'cypress';
      import { nxComponentTestingPreset } from '@nx/next/plugins/component-testing';

      export default defineConfig({
        component: nxComponentTestingPreset(__filename),
      });
      "
    `);
    expect(tree.exists('libs/demo/pages/index.cy.ts')).toBeFalsy();
    expect(
      readProjectConfiguration(tree, 'demo').targets['component-test']
    ).toEqual({
      executor: '@nx/cypress:cypress',
      options: {
        cypressConfig: 'libs/demo/cypress.config.ts',
        skipServe: true,
        testingType: 'component',
      },
    });
  });
});
