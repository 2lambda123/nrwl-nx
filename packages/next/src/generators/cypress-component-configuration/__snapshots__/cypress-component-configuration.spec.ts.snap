// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cypress-component-configuration generator should setup nextjs app 1`] = `
{
  "compilerOptions": {
    "module": "commonjs",
    "outDir": "../../dist/out-tsc",
    "types": [
      "cypress",
      "node",
    ],
  },
  "extends": "../tsconfig.json",
  "include": [
    "support/**/*.ts",
    "../cypress.config.ts",
    "../**/*.cy.ts",
    "../**/*.cy.tsx",
    "../**/*.cy.js",
    "../**/*.cy.jsx",
    "../**/*.d.ts",
  ],
}
`;

exports[`cypress-component-configuration generator should setup nextjs lib 1`] = `
{
  "compilerOptions": {
    "module": "commonjs",
    "outDir": "../../dist/out-tsc",
    "types": [
      "cypress",
      "node",
    ],
  },
  "extends": "../tsconfig.json",
  "include": [
    "support/**/*.ts",
    "../cypress.config.ts",
    "../**/*.cy.ts",
    "../**/*.cy.tsx",
    "../**/*.cy.js",
    "../**/*.cy.jsx",
    "../**/*.d.ts",
  ],
}
`;
